src/credential/credentialController.ts

```diff
@@ -0,0 +1,219 @@
+import * as vscode from 'vscode';
+import { AzureEnvironment } from 'ms-rest-azure';
+import * as template from 'url-template';
+import { UserInfo, DocsSignInStatus, EXTENSION_ID, uriHandler, extensionConfig } from '../shared';
+import { parseQuery, delay, openUrl } from '../utils/utils';
+import { UserSigningIn, UserSignInSucceeded, UserSignedOut, CredentialReset, UserSignInFailed, BaseEvent, SignInProgress, CredentialRetrieveFromLocalCredentialManager } from '../common/loggingEvents';
+import { EventType } from '../common/eventType';
+import { EventStream } from '../common/eventStream';
+import { KeyChain } from './keyChain';
+import { EnvironmentController } from '../common/environmentController';
+
+async function handleAuthCallback(callback: (uri: vscode.Uri, resolve: (result: any) => void, reject: (reason: any) => void) => void): Promise<any> {
+    let uriEventListener: vscode.Disposable;
+    return Promise.race([
+        delay(extensionConfig.SignInTimeOut, new Error(`Timeout`)),
+        new Promise((resolve: (result: any) => void, reject: (reason: any) => void) => {
+            uriEventListener = uriHandler.event((uri) => callback(uri, resolve, reject));
+        }).then(result => {
+            uriEventListener.dispose();
+            return result;
+        }).catch(err => {
+            uriEventListener.dispose();
+            throw err;
+        })
+    ]);
+}
+
+export interface Credential {
+    readonly signInStatus: DocsSignInStatus;
+    readonly aadInfo: string;
+    readonly userInfo: UserInfo;
+}
+
+export class CredentialController implements vscode.Disposable {
+    private signInStatus: DocsSignInStatus;
+    private aadInfo: string | undefined;
+    private userInfo: UserInfo | undefined;
+
+    public onDidChangeCredential: vscode.Event<Credential>;
+
+    constructor(private keyChain: KeyChain, private eventStream: EventStream, private environmentController: EnvironmentController) { }
+
+    public dispose(): void {

```

Comment:
>dispose [](start = 11, length = 7)

why empty `dispose`?

Reply:
can be removed, nothing need to be disposed

---
In reply to: [363593823](https://github.com/docascode/vscode-docs-build/pull/10#discussion_r363593823) [](ancestors = 363593823)